util/txesbmantool/txesbmantool.c

```diff
@@ -0,0 +1,1461 @@
+
+	ret = wc_RsaFlattenPublicKey(&rsaKey, (byte *)&key_struct->exponent, &e_size,
+					      (byte *)key_struct->modulus, &n_size);
+
+	if (ret != 0) {
+		ERROR("Failed to write key modulus and exponent! %d\n", ret);
+		goto exit;
+	}
+
+	/* Change endianess of the modulus to LE */
+	if (swap_bytes((byte *)key_struct->modulus, n_size)) {
+		ERROR("Failed to change RSA Key modulus endianess!\n");
+		goto exit;
+	}
+
+	if (n_size * 8 != RSA_KEY_SIZE) {
+		ERROR("Incorrect RSA key size! %d != 2048\n", n_size * 8);
+		ret = EXIT_FAILURE;
+		goto exit;
+	}
+
+	if (key_struct->exponent != RSA_KEY_EXPONENT) {
+		ERROR("Incorrect RSA key size! 0x%x != 0x10001\n", key_struct->exponent);
+		ret = EXIT_FAILURE;
+		goto exit;
+	}
+
+	/* Verify hash type is supported */
+	if (wc_HashGetDigestSize(WC_HASH_TYPE_SHA256) != 32) {
+		ERROR("Hash type SHA256 not supported!\n");
+		ret = EXIT_FAILURE;
+		goto exit;
+	}
+
+	sigLen = wc_SignatureGetSize(WC_SIGNATURE_TYPE_RSA_W_ENC, &rsaKey, sizeof(rsaKey));
+	if (sigLen != sizeof(key_struct->signature)) {
+		ERROR("RSA Signature size check fail! %d\n", sigLen);
+		ret = EXIT_FAILURE;
+		goto exit;
+	}
+
+	sigBuf = malloc(sigLen);
+	if (!sigBuf) {
+		ERROR("RSA Signature malloc failed!\n");
+		ret = EXIT_FAILURE;
+		goto exit;
+	}
+
+	ret = wc_SignatureGenerate(WC_HASH_TYPE_SHA256, WC_SIGNATURE_TYPE_RSA_W_ENC,
+				   (byte *)man,
+				   MANIFEST_SIZE - sizeof(struct key_signature),
+				   sigBuf, &sigLen,
+				   &rsaKey, sizeof(rsaKey),
+				   &rng);
+	if (ret < 0) {
+		ERROR("RSA signature generation failed (%d)!\n", ret);
+		goto exit;
+	}
+
+	ret = wc_SignatureVerify(WC_HASH_TYPE_SHA256, WC_SIGNATURE_TYPE_RSA_W_ENC,
+				 (byte *)man,
+				 MANIFEST_SIZE - sizeof(struct key_signature),
+				 sigBuf, sigLen,
+				 &rsaKey, sizeof(rsaKey));
+	if (ret < 0) {
+		ERROR("RSA Signature Verification failed: (%d)\n", ret);
+		goto exit;
+	}
+
+	if (swap_bytes(sigBuf, sigLen)) {
+		ERROR("Failed to change RSA Signature endianess!\n");
+		goto exit;
+	}
+
+	memcpy(key_struct->signature, sigBuf, sigLen);
+exit:
+	if(sigBuf) {
+		free(sigBuf);
+	}
+	wc_FreeRsaKey(&rsaKey);
+	wc_FreeRng(&rng);
+
+	return ret;
+}
+
+static int save_buffer_to_file(const char *filename, byte *buffer, word32 bufLen)
+{
+	int ret = 0;
+	FILE* file = NULL;
+
+	file = fopen(filename, "wb");
+	if (file == NULL) {
+		ERROR("Failed to create file %s!\n", filename);
+		ret = EXIT_FAILURE;
+		goto exit;
+	}
+
+	if(fwrite(buffer, 1, bufLen, file) != bufLen) {
+		ERROR("Error writing %s file! %d", filename, ret);
+		ret = EXIT_FAILURE;
+		goto exit;
+	}
+
+exit:
+	if(file) {
+		fclose(file);
+	}
+
+	return ret;
+}
+
+static int load_file_to_buffer(const char *filename, byte **fileBuf, word32 *fileLen)
+{
+	int ret = 0;
+	FILE *file = NULL;
+
+	file = fopen(filename, "rb");
+	if (file == NULL) {
+		ERROR("File %s does not exist!\n", filename);
+		ret = EXIT_FAILURE;
+		goto exit;
+	}
+
+	*fileLen = get_file_size(file);
```

Comment:
https://github.com/Dasharo/coreboot/pull/485/files#r1638186901, feel free to improve or re-wrap  the message:

```suggestion
	*fileLen = get_file_size(file);

	/*
	 * To avoid undefined behavior when reading from potentially unaligned
	 * pointer, check that the file size is divisible by 4 bytes. Both manifests
	 * have such sizes (see SB_MANIFEST_SIZE and KEY_MANIFEST_SIZE), so this
	 * wouldn't fail for valid manifests. In case of coreboot.rom, the offsets are
	 * calculated relatively to the end of file, so the result may be misaligned if
	 * file size is misaligned. Instead of repeating checks for proper alignment in
	 * cmd_verify() and cmd_print(), do it here.
	 */
	if (*fileLen % sizeof(uint32_t)) {
		ERROR("File size isn't a multiple of 4 bytes!\n");
		ret = EXIT_FAILURE;
		goto exit;
	}
```